### PR TITLE
factored out command logic from the cli

### PR DESCRIPTION
**Pull request by substack**
This patch provides a set of default commands taken from the command-line client. With this change, we can roll out new commands to all of the clients.

### CODE REVIEW

File: src/client.js [hunk 1]
```diff
@@ -228,6 +240,50 @@ class Client {
...
 
+  /**
+   * Add a command to the set of supported commands.
+   * @param {string} [name] the long-form command name
+   * @param {object} [cmd] the command object
+   * @param {function} [cmd.help] function returning help text
+   * @param {array} [cmd.alias] array of string aliases
+   * @param {function} [cmd.call] implementation of the command receiving (cabal, res, arg) arguments
+   */
+  addCommand(name, cmd) {
```
Comment (cblgh) [inline src/client.js:251]: what's the intent for this function? it doesn't seem to be used in cabal-client or in the cli refactor

File: src/cabal-details.js [hunk 2]
```diff
@@ -1,6 +1,7 @@
 const EventEmitter = require('events')
 const debug = require("debug")("cabal-client")
 const { VirtualChannelDetails, ChannelDetails } = require("./channel-details")
+const { nextTick } = process
```
Comment (cblgh) [inline src/cabal-details.js:4]: unused?

File: src/cabal-details.js [hunk 3]
```diff
@@ -123,9 +174,12 @@ class CabalDetails extends EventEmitter {
...
    */
   publishNick(nick, cb) {
-    this._cabal.publishNick(nick, cb)
-    this.user.name = nick
-    this._emitUpdate("publish-nick", { name: nick })
+    this._cabal.publishNick(nick, (err) => {
+      if (err) return cb(err)
+      this.user.name = nick
+      this._emitUpdate("publish-nick", { name: nick })
+      cb()
```
Comment (cblgh) [inline src/cabal-details.js:181]: nice catch

File: src/client.js [hunk 4]
```diff
@@ -41,6 +42,13 @@ class Client {
     this.currentCabal = null
     this.config = opts.config
     this.maxFeeds = opts.maxFeeds || 1000
+    this.aliases = opts.aliases || {}
+    this.commands = Object.assign({}, defaultCommands, opts.commands)
```
Comment (cblgh) [inline src/client.js:46]: ahh nice - this should e.g. allow for cabal-desktop to overload any default commands, right?

File: src/client.js (hunk 1 above)
Comment (substack) [reply src/client.js:251]: so you can adjust the set of commands at runtime. for things like hot-loading plugins for example

File: src/client.js (hunk 4 above)
Comment (substack) [reply src/client.js:46]: yes. you can also use client.{add,remove}Command() and client.{add,remove}Alias() methods

Comment (cblgh) [reply src/client.js:46]: cc @nikolaiwarner good to know for desktop ^ :3

**Pull request merged**